docs/design/extending-api.md

```diff
@@ -0,0 +1,169 @@
+# Extending the Kubernetes API
+
+This document describes the design for implementing custom API types in the Kubernetes API Server.
+
+
+## Object Model
+
+### The Schema Object
+The `Schema` object describes the object schema(s) of multiple versions of a custom object that the user wants to add
+to the Kubernetes API.
+
+Each Schema object has the standard Kubernetes metadata.
+
+Additionally it has an array of one more `Version` objects.
+
+### The Version Object
+The `Version` object describes a single concrete version of a custom object.  The `Version` object has
+two fields:
+   * `Name` which is the name of the version (e.g. `v1beta3`)
+   * `Spec` which is the specification of the version
+
+### The SchemaSpec Object
+The `SchemaSpec` object describes a concrete schema for an object, it consists of one or more `Field` objects.
+
+
+### The Field Object
+Each `Field` object has two fields:
+   * `Name` which is the name of the field (e.g. `color`)
+   * `Type` which describes the type of the field (e.g. `int`)
+
+### The Type Object
+Each `Type` object has three fields:
+   * `Kind` which specifies the kind of the field.  The possible values correspond to JSON types and are:
+      * `Number`
+      * `String`
+      * `Boolean`
+      * `Array`
+      * `Object`
+   * `Schema` is a required field if the `Kind` is `Object` and may not be specified if `Kind` is not `Object`.
```

Comment:
is Schema a *Type field (recursive) or a name of another Schema object?

Assuming it is recursive, prefer ObjectSchema to make it clear it goes with Kind=Object?